pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,618 @@
+                //append aborted txn nex time.
+                log.error("Failed to take snapshot segment [{}] at maxReadPosition [{}] "
+                                + "for the topic [{}], and the size of the segment is [{}]",
+                        this.sequenceID, abortedMarkerPersistentPosition, topic.getName(), sealedAbortedTxnIdSegment.size(), e);
+                return null;
+            });
+        }
+
+        private CompletableFuture<Void> writeSnapshotSegmentAsync(LinkedList<TxnID> segment,
+                                                                  PositionImpl abortedMarkerPersistentPosition) {
+            TransactionBufferSnapshotSegment transactionBufferSnapshotSegment = new TransactionBufferSnapshotSegment();
+            transactionBufferSnapshotSegment.setAborts(convertTypeToTxnIDData(segment));
+            transactionBufferSnapshotSegment.setTopicName(this.topic.getName());
+            transactionBufferSnapshotSegment.setPersistentPositionEntryId(abortedMarkerPersistentPosition.getEntryId());
+            transactionBufferSnapshotSegment.setPersistentPositionLedgerId(abortedMarkerPersistentPosition.getLedgerId());
+
+            return snapshotSegmentsWriterFuture.thenCompose(segmentWriter -> {
+                transactionBufferSnapshotSegment.setSequenceId(this.sequenceID.get());
+                return segmentWriter.writeAsync(buildKey(this.sequenceID.get()), transactionBufferSnapshotSegment);
+            }).thenCompose((messageId) -> {
+                PositionImpl maxReadPosition = topic.getMaxReadPosition();
+                //Build index for this segment
+                TransactionBufferSnapshotIndex index = new TransactionBufferSnapshotIndex();
+                index.setSequenceID(transactionBufferSnapshotSegment.getSequenceId());
+                index.setMaxReadPositionLedgerID(maxReadPosition.getLedgerId());
+                index.setMaxReadPositionEntryID(maxReadPosition.getEntryId());
+                index.setPersistentPositionLedgerID(((MessageIdImpl) messageId).getLedgerId());
+                index.setPersistentPositionEntryID(((MessageIdImpl) messageId).getEntryId());
+
+                indexes.put(maxReadPosition, index);
+                //update snapshot segment index.
+                return updateSnapshotIndex(new TransactionBufferSnapshotIndexesMetadata(
+                                maxReadPosition.getLedgerId(), maxReadPosition.getEntryId(), new LinkedList<>()),
+                        indexes.values().stream().toList());
+            });
+        }
+
+        private CompletableFuture<Void> deleteSnapshotSegment(List<PositionImpl> positionNeedToDeletes) {
+            List<CompletableFuture<Void>> results = new ArrayList<>();
+            for (PositionImpl positionNeedToDelete : positionNeedToDeletes) {
+                long sequenceIdNeedToDelete = indexes.get(positionNeedToDelete).getSequenceID();
```

Comment:
the positionNeedToDeletes is abortMarkPersistentPostion not maxReadPosition how we can get the correct segement sequenceId

Reply:
Good catch! I think we should change the key of the indexes to `abortMarkPersistentPostion`.